### PR TITLE
Refactor terraform to support 0.12, and bump k8s to v1.14.6

### PR DESCRIPTION
**Pull request by smalltown**
- All: support terraform 0.12 (#32)
- All: Kubernetes upgrade to 1.14.6
- ElastiKube: Etcd upgrade to v3.4.0, CoreDNS upgrade to 1.6.2 , flannel upgrade to 0.11.0

### CODE REVIEW

**Comment by smalltown** [comment]
#32

**Comment by pragmaticivan** [comment]
Just created a cluster with that PR, runs great. Thanks!

**Comment by smalltown** [comment]
Wow ! @pragmaticivan thank for your feedback very much, 
but I only push the ESK related commit to this PR,
I still cooking the self-hosted part,
and the release will be blocked by one third party provider Ignition [PR](https://github.com/terraform-providers/terraform-provider-ignition/pull/56)

**Comment by smalltown** [comment]
Done of the development, but not complete yet, need to 
1) wait for third party provider (ignition) issue fixed
2) modify the document
3) code review

File: modules/ignitions/kube-config/kubeconfig.tf [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
 
 data "ignition_file" "kubeconfig" {
   filesystem = "root"
   path       = "/etc/kubernetes/kubeconfig"
-  mode       = 0644
+  mode       = 420
 
   content {
-    content = "${coalesce(var.content, join("", data.template_file.kubeconfig.*.rendered))}"
+    content = coalesce(var.content, join("", data.template_file.kubeconfig.*.rendered))
```
Comment (alanchchen) [inline modules/ignitions/kube-config/kubeconfig.tf]: Can we remove this `join` ?

File: modules/ignitions/kube-config/outputs.tf [hunk 2]
```diff
@@ -4,10 +4,10 @@ output "systemd_units" {
...
 output "files" {
   value = [
-    "${data.ignition_file.kubeconfig.id}",
+    data.ignition_file.kubeconfig.rendered,
   ]
 }
 
 output "content" {
-  value = "${coalesce(var.content, join("", data.template_file.kubeconfig.*.rendered))}"
+  value = coalesce(var.content, join("", data.template_file.kubeconfig.*.rendered))
```
Comment (alanchchen) [inline modules/ignitions/kube-config/outputs.tf]: ditto

File: modules/ignitions/etcd/certs.tf [hunk 3]
```diff
@@ -1,83 +1,83 @@
...
 }
 
 data "ignition_file" "etcd_server_cert" {
   path       = "${var.certs_path}/server.crt"
-  mode       = 0400
-  uid        = 232
-  gid        = 232
+  mode       = 256
+  uid        = var.etcd_user_id
+  gid        = var.etcd_user_id
```
[the remark this answers is not in the document]
Comment (smalltown) [reply modules/ignitions/etcd/certs.tf]: sure, let me add it in next commit

File: modules/ignitions/kube-config/kubeconfig.tf (hunk 1 above)
Comment (smalltown) [reply modules/ignitions/kube-config/kubeconfig.tf]: let me check it, thank for reviewing

Comment (smalltown) [reply modules/ignitions/kube-config/kubeconfig.tf]: if disable user to pass their own kubeconfig content, then this join can be disabled

**Pull request merged**